hail/src/main/scala/is/hail/expr/ir/Simplify.scala

```diff
@@ -221,6 +221,8 @@ object Simplify {
 
     case ArrayFilter(ArraySort(a, left, right, compare), name, cond) => ArraySort(ArrayFilter(a, name, cond), left, right, compare)
 
+    case ArrayFilter(ToStream(ArraySort(a, left, right, compare)), name, cond) => ToStream(ArraySort(ArrayFilter(a, name, cond), left, right, compare))
```

Comment:
let's also add rules:
```
case ToArray(ToStream(a)) if a.typ.isInstanceOf[TArray] => a
case ToStream(ToArray(s)) if s.typ.isInstanceOf[TStream] => s
case ToStream(Let(name, value, ToArray(x)) if x.typ.isInstanceOf[TStream] => Let(name, value, x)
```

Reply:
Also ToStream(ArrayMap(ToStream(a)) and same for ArrayFold, ArrayFold2 right. ArrayZip could use one as well (ToStream(ArrayZip(x)) if x.typ.isInstanceOf[TStreamable]). Not sure how far we want to take that

On the other hand, should our lowering pass catch these items? I don't think we should be returning ToStream(ToArray)...

Reply:
this one seems wrong: `ToStream(ArrayMap(ToStream(a))`

the outer ToStream is invalid since the ArrayMap is already of type stream.

If there's any additional complexity to handling this in lowering, it's not worth it. Composition of simple, modular pieces is a better design than complexity loaded into one place.